src/common/transformations/src/transformations/common_optimizations/compress_float_constants.cpp

```diff
@@ -25,19 +28,43 @@ std::shared_ptr<ov::Node> change_constant_precision_to_fp16(std::shared_ptr<ov::
     if (dst_data == nullptr)
         return nullptr;
 
-    bool is_overflow = false;
+    bool is_out_of_range = false;
+    int num_out_of_range = 0;
     for (size_t i = 0; i < size; ++i) {
-        if (src_data[i] > std::numeric_limits<ov::float16>::max()) {
+        // if it is smaller than the smallest positive normal fp16 but not zero
+        if (std::abs(src_data[i]) <= ov::float16::from_bits(0x0400) && src_data[i] != 0.0f) {
+            num_out_of_range++;
+            is_out_of_range = true;
+        } else if (src_data[i] > std::numeric_limits<ov::float16>::max()) {
             dst_data[i] = std::numeric_limits<ov::float16>::max();
-            is_overflow = true;
+            is_out_of_range = true;
+            num_out_of_range++;
         } else if (src_data[i] < std::numeric_limits<ov::float16>::lowest()) {
             dst_data[i] = std::numeric_limits<ov::float16>::lowest();
-            is_overflow = true;
+            is_out_of_range = true;
+            num_out_of_range++;
         } else {
             dst_data[i] = static_cast<ov::float16>(src_data[i]);
         }
     }
-    if (is_overflow) {
+
+    // thresholds in percents
+    // if more than 75% of a FP32 constant do not fit into FP16 keep in FP32
+    auto keep_threshold = static_cast<float>(ov::util::getenv_int("KEEP_FP32_THRESHOLD", 75));
+    auto actual_out_of_range_in_percent = 100.0f * static_cast<float>(num_out_of_range) / static_cast<float>(size);
+
+    if (actual_out_of_range_in_percent >= keep_threshold) {
+        std::cerr << "Warning: while the model was compressed into FP16, Constant '" << constant->get_friendly_name()
+                  << "' "
+                     "is kept in FP32. This means the model originally was not designed to be inferred on FP16. "
+                     "Therefore inference results on FP16 supporting plugins potentially can significantly "
+                     "differ from FP32."
+                  << std::endl;
```

Comment:
please use OPENVINO routines

Reply:
I tried to replace it with `NGRAPH_WARN` but in that case this log is not showed in MO outputs. I think it's important to notice user if network data are not fit into fp16 it's a sign that potentially network wouldn't work correctly. Even though in most cases we get adequate accuracy for compressed_to_fp16 networks in general case we cannot guarantee that such networks will work correctly and if used face such situation it's better if he was informed.

Attached log example for MO
![image](https://user-images.githubusercontent.com/5703039/213567391-fd9c6c07-1f70-4ccb-9b05-6091337a3b34.png)

Reply:
Can we please merge as is and discuss warning after? Seems like at the moment is more important to see how solution works in general and get some validation reports. Details we can discuss after merge

Seems we more discuss details not the basiс solution. @slyalin do you agree to merge as is?

Reply:
I don't think that user can really decide which model `was not designed to be inferred on FP16`, so such a message will only bring some confusion. Our method of sticking to FP32 is designed to avoid manual control from user side. Can you show me at least one example where such message is really valuable.

From the other hand, if we suppose that such a message should help the user, I don't see any way of control over the level of diagnostics for such messages. In mo as a tool, we have this debug level command line parameter, in the runtime we don't have it. Always printing such messages is not really a good idea, but I don't have any other options (again if it is really useful).

Reply:
Hmm, indeed, i must admit i don't have example models where this message could be valuable.

And yes, we don't have control over this dubug level prints. Therefore i removed this prints.